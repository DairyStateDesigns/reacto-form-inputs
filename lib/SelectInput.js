import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';

import customPropTypes from './shared/propTypes';

class SelectInput extends Component {
  static isFormInput = true;

  static propTypes = {
    ...customPropTypes.inputs,
    className: PropTypes.string,
    emptyLabel: PropTypes.string,
    options: customPropTypes.options,
    style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool,
    ]),
  };

  static defaultProps = {
    className: undefined,
    emptyLabel: '(Select One)',
    isReadOnly: false,
    name: undefined,
    onChange() {},
    onChanging() {},
    options: [],
    placeholder: undefined,
    style: {},
    value: undefined,
  };

  constructor(props) {
    super(props);

    this.validateOptions(props.options);

    this.state = {
      value: props.value || null,
    };
  }

  componentWillMount() {
    this.handleChanged(this.state.value);
  }

  componentWillReceiveProps(nextProps) {
    const { options, value } = this.props;
    const { options: nextOptions, value: nextValue } = nextProps;

    // Whenever a changed value prop comes in, we reset state to that, thus becoming clean.
    if (!isEqual(value, nextValue)) {
      this.setValue(nextValue);
    }

    if (!isEqual(options, nextOptions)) {
      this.validateOptions(nextOptions);
    }
  }

  onChange = (event) => {
    let { value } = event.target;

    if (value !== undefined && value !== null) {
      switch (this.dataType) {
        case 'string':
          value = String(value);
          break;
        case 'number':
          value = value === '' ? null : Number(value);
          break;
        case 'boolean':
          value = value === '' ? null : Boolean(value);
          break;
        default:
          // do nothing
      }
    }

    this.setValue(value);
  };

  getValue() {
    return this.state.value;
  }

  setValue(value) {
    this.setState({ value });
    this.handleChanged(value);
  }

  resetValue() {
    this.setValue(this.props.value);
  }

  handleChanged(value) {
    const { onChange, onChanging } = this.props;
    if (value !== this.lastValue) {
      this.lastValue = value;
      onChanging(value);
      onChange(value);
    }
  }

  // Input is dirty if value prop doesn't match value state. Whenever a changed
  // value prop comes in, we reset state to that, thus becoming clean.
  isDirty() {
    return !isEqual(this.state.value, this.props.value);
  }

  // Make sure all option values have the same data type, and record what that is
  validateOptions(options) {
    (options || []).forEach((option) => {
      if (option.optgroup) {
        this.validateOptions(option.options);
      } else {
        const checkDataType = typeof option.value;
        if (!this.dataType) {
          this.dataType = checkDataType;
        } else if (checkDataType !== this.dataType) {
          throw new Error(`reacto-form SelectInput: All option values must have the same data type. The data type of the first option is "${this.dataType}" while the data type of the ${option.label} option is "${checkDataType}"`);
        }
      }
    });
  }

  renderOptions() {
    const { emptyLabel, options } = this.props;

    const userOpts = (options || []).map((option) => {
      if (option.optgroup) {
        return (
          <optgroup key={option.id || `${option.value}`} label={option.optgroup}>
            {(options.options || []).map((opt) => {
              return (
                <option key={opt.id || `${opt.value}`} value={opt.value}>{opt.label}</option>
              );
            })}
          </optgroup>
        );
      }
      return (
        <option key={option.id || `${option.value}`} value={option.value}>{option.label}</option>
      );
    });

    return [<option key="reacto-form-default" value="">{emptyLabel}</option>].concat(userOpts);
  }

  render() {
    const { className, isReadOnly, name, style } = this.props;
    const { value } = this.state;

    return (
      <select
        className={className}
        readOnly={isReadOnly}
        name={name}
        onChange={this.onChange}
        style={style}
        value={value || ''}
      >
        {this.renderOptions()}
      </select>
    );
  }
}

export default SelectInput;
