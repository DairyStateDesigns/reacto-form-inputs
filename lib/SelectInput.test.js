import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { testInput } from 'composable-form-tests';
import SelectInput from './SelectInput';

const stringOptions = [
  { label: 'A', value: 'a' },
  { label: 'B', value: 'b' },
  { label: 'C', value: 'c' },
];

const numberOptions = [
  { label: '1', value: 1 },
  { label: '2', value: 2 },
  { label: '3', value: 3 },
];

const booleanOptions = [
  { label: 'Yes', value: true },
  { label: 'No', value: false },
];

// Run generic Input tests that should pass for any Composable Forms Spec input
testInput({
  component: SelectInput,
  defaultValue: null,
  exampleValueOne: 'a',
  exampleValueTwo: 'b',
  mount,
  options: stringOptions,
  simulateChanged(wrapper, value) {
    wrapper.find('select').simulate('change', { target: { value } });
  },
});

// Test with number options
testInput({
  component: SelectInput,
  defaultValue: null,
  exampleValueOne: 1,
  exampleValueTwo: 2,
  mount,
  options: numberOptions,
  simulateChanged(wrapper, value) {
    wrapper.find('select').simulate('change', { target: { value } });
  },
});

// Test with boolean options
testInput({
  component: SelectInput,
  defaultValue: null,
  exampleValueOne: true,
  exampleValueTwo: false,
  mount,
  options: booleanOptions,
  simulateChanged(wrapper, value) {
    wrapper.find('select').simulate('change', { target: { value } });
  },
});

test('renders', () => {
  const component = renderer.create(
    <SelectInput name="test" options={stringOptions} />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
