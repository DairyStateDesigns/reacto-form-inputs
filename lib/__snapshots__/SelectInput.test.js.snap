// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<select
  name="test"
  onChange={[Function]}
  readOnly={false}
  style={Object {}}
  value=""
>
  <option
    value=""
  >
    (Select One)
  </option>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
  <option
    value="c"
  >
    C
  </option>
</select>
`;
